src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1014 @@
+
+    wc->rclient.x0 = rx0;
+    wc->rclient.x1 = rx1;
+    wc->rclient.y0 = ry0;
+    wc->rclient.y1 = ry1;
+    wc->rclient.w = rx1 - rx0;
+    wc->rclient.h = ry1 - ry0;
+
+    wc->style = style;
+    wc->exstyle = exstyle;
+    wc->spwnd_addr = win;
+    wc->text = wn;
+    wc->atom = atom;
+
+    return wc;
+}
+
+bool filter_wnd_text(wchar_t* c)
+{
+    /* Checks, if a target text is present */
+    for (size_t j = 0; j < ARRAY_SIZE(BTN_TEXTS); j++)
+    {
+        if (c)
+        {
+            const wchar_t* m = wcswcs(c, BTN_TEXTS[j]);
+            if (!m)
+                continue;
+
+            if (m - c < BTN_THRESHOLD)
+            {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+int find_button_to_click(vmi_instance_t vmi, struct desktop* desk,
+    GArray* wins, struct wnd* btn)
+{
+    int res = 0;
+    /* Current window address */
+    addr_t wa = 0;
+    /* Current window parsed container */
+    struct wnd* wnd = NULL;
+    /* Root window used for size calculations */
+    struct wnd* root = NULL;
+    /* Candidate window */
+    struct wnd* cand = NULL;
+    /* Visibile part of candidate */
+    struct rect* r = NULL;
+    /* Maximum button size to consider */
+    uint16_t mw, mh;
+    /* Signifies a matching wnd text */
+    bool match = false;
+
+    if (!wins)
+        return -1;
+
+    /* Gets the desktop pane  */
+    wa = g_array_index(wins, addr_t, 0);
+
+    root = construct_wnd_container(vmi, desk->providing_pid, wa);
+    /*
+     * Naive assumption, that buttons will be at least BTN_RATIO times smaller
+     * than the respective desktop dimension
+     */
+    mw = root->r.x1 / BTN_RATIO;
+    mh = root->r.y1 / BTN_RATIO;
+
+    /* Frame of desktop */
+    size_t w = root->r.x1;
+    size_t h =root->r.y1;
+    free_wnd_container(root);
+
+    /* Keeping track of occupied screen locations with a bitmap */
+    size_t n = ((w + w % 8) / 8) * h;
+    char* map = (char*) calloc(1, sizeof(char) * n);
+    if (!map)
+    {
+        printf("[HIDSIM][MONITOR] Memory allocation pixel map failed\n");
+        return -1;
+    }
+
+    size_t l = wins->len;
+
+    for (size_t i = 0; i < l; i++)
+    {
+        wa = g_array_index(wins, addr_t, l - (i+1));
+        wnd = construct_wnd_container(vmi, desk->providing_pid, wa);
+
+        /* Performs filtering based on size */
+        if (wnd->r.w > mw || wnd->r.h > mh)
+        {
+            update_visibility_bitmask(map, n, w, &wnd->r);
+            free_wnd_container(wnd);
+            continue;
+        }
+        if (wnd->text)
+        {
+            match = filter_wnd_text(wnd->text);
+            if (match)
+                cand = wnd;
+        }
+#ifndef DISABLE_ATOMS
+        /* Refines selection by filtering based on window class */
+        if (cand && cand->atom )
+        {
+            struct atom_entry* a = (struct atom_entry*) g_hash_table_lookup(
+                    desk->atom_table, GUINT_TO_POINTER(wnd->atom));
```

Comment:
shouldn't we check returning value of g_hash_table_lookup here and in other similar places?